### PR TITLE
Reset the Mouse Cursor While Selecting

### PR DESCRIPTION
**Pull request by nixpulvis**
This change makes the mouse cursor reset to it's default state while a selection is in progress. A click to clear the selection doesn't trigger a URL action, and so the mouse hand style isn't applied until the selection is cleared.

Before:
before

After:
after

### CODE REVIEW

File: alacritty_terminal/src/input.rs [hunk 1]
```diff
@@ -455,7 +455,11 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
             if let Some(url) =
                 self.ctx.terminal().urls().drain(..).find(|url| url.contains(buffer_point))
             {
-                return MouseState::Url(url);
+                if self.ctx.selection_is_empty() &&
+                   self.ctx.mouse().left_button_state != ElementState::Pressed
```
Comment (chrisduerr) [inline alacritty_terminal/src/input.rs]: Based on my testing, this shouldn't be necessary. I've asked about this before, but could you go into more detail why this is needed?

File: alacritty_terminal/src/input.rs [hunk 2]
```diff
@@ -455,7 +455,11 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
             if let Some(url) =
                 self.ctx.terminal().urls().drain(..).find(|url| url.contains(buffer_point))
             {
-                return MouseState::Url(url);
+                if self.ctx.selection_is_empty() &&
```
Comment (chrisduerr) [inline alacritty_terminal/src/input.rs]: I think we can move this check to the other `if` block, the one before the `if let` binding. We don't need access to the `url` so it should be easier to check it early in one block.

File: alacritty_terminal/src/input.rs [hunk 3]
```diff
@@ -692,6 +696,17 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
...
+            let mouse_state = self.mouse_state(point);
+            self.update_mouse_cursor(mouse_state);
+            match mouse_state {
+                MouseState::Url(url) => {
+                    let url_bounds = url.linear_bounds(self.ctx.terminal());
+                    self.ctx.terminal_mut().set_url_highlight(url_bounds);
+                },
+                _ => {
+                    self.ctx.terminal_mut().reset_url_highlight();
+                }
```
[the remark this answers is not in the document]
Comment (chrisduerr) [reply alacritty_terminal/src/input.rs]: Why would we need this? This code is just to make sure that when the selection is cleared by clicking, that URLs are shown again, right? We're never concerned about removing URLs when the selection is cleared (since it shouldn't exist in the first place).

So this can just be transformed into an `if let`.

File: CHANGELOG.md [hunk 4]
```diff
@@ -42,6 +42,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Intermediate bytes of CSI sequences not checked
 - Wayland clipboard integration
 - Use text mouse cursor when mouse mode is temporarily disabled with shift
+- Remove URL styles when selecting text
```
Comment (chrisduerr) [inline CHANGELOG.md]: I wouldn't know what `URL styles` are and I'd assume I'm not alone here. I think we can describe better what we're doing here.

Not much better, but maybe this would work:

> - Block URL highlight while a selection is active

Open for other suggestions though. I just don't think `Remove URL styles` says a lot.

File: CHANGELOG.md [hunk 5]
```diff
@@ -42,6 +42,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Intermediate bytes of CSI sequences not checked
 - Wayland clipboard integration
 - Use text mouse cursor when mouse mode is temporarily disabled with shift
+- Block URL highlight while a selection is active.
```
Comment (chrisduerr) [inline CHANGELOG.md]: No punctuation at the end here.

File: alacritty_terminal/src/input.rs [hunk 6]
```diff
@@ -692,6 +694,12 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
             self.mouse_report(code, ElementState::Released, modifiers);
             return;
         } else if let (Some(point), true) = (point, button == MouseButton::Left) {
+            let mouse_state = self.mouse_state(point);
+            self.update_mouse_cursor(mouse_state);
+            if let MouseState::Url(url) = mouse_state {
+                let url_bounds = url.linear_bounds(self.ctx.terminal());
+                self.ctx.terminal_mut().set_url_highlight(url_bounds);
+            }
```
Comment (chrisduerr) [inline alacritty_terminal/src/input.rs]: Why is this required here? The mouse highlight is accurate for me after clicking on URLs even without this.

Comment (chrisduerr) [reply alacritty_terminal/src/input.rs]: Ah, got it!

**Comment by nixpulvis** [comment]
Fixed the rebase, and squashed down to a single commit.

File: alacritty_terminal/src/input.rs (hunk 6 above)
Comment (chrisduerr) [reply alacritty_terminal/src/input.rs]: This makes me think though, couldn't we move the `self.launch_url()` call into this then and remove the check for modifiers at the start of the `launch_url` method?

**Comment by nixpulvis** [comment]
Note to self, from @chrisduerr: https://paste.rs/YUx.diff

**Comment by nixpulvis** [comment]
I've uncovered a bug while testing this, however this is actually a bug on master, so I don't think it should block merging this.

block-selection-url-bug-2019-09-07_00 03 52

To reproduce on master, create a 0 width N height block selection, and then attempt to click a URL. I believe a proper fix for this is to disallow 0 width block selections, similarly to how other selections disappear.

**Comment by chrisduerr** [comment]
Very good catch @nixpulvis. I'll see if I can find a simple patch, but I do agree that it should be unrelated to this PR.

File: alacritty_terminal/src/input.rs [hunk 7]
```diff
@@ -691,34 +693,34 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
...
         }
 
-        let point = self.ctx.terminal().visible_to_buffer(point);
-        let url = self.ctx.terminal().urls().drain(..).find(|url| url.contains(point))?;
-        let text = self.ctx.terminal().url_to_string(&url);
-
-        let launcher = self.mouse_config.url.launcher.as_ref()?;
-        let mut args = launcher.args().to_vec();
-        args.push(text);
+        let text = self.ctx.terminal().url_to_string(url);
```
Comment (chrisduerr) [inline alacritty_terminal/src/input.rs]: This is useless if we don't have url launcher set. So instead we can do it later inside the `if let`.
